src/kudu/consensus/routing.cc

```diff
@@ -0,0 +1,488 @@
+namespace consensus {
+
+////////////////////////////////////////////////////////////////////////////////
+// RoutingTable
+////////////////////////////////////////////////////////////////////////////////
+
+Status RoutingTable::Init(const RaftConfigPB& raft_config,
+                          const ProxyTopologyPB& proxy_topology,
+                          const std::string& leader_uuid) {
+  unordered_map<string, Node*> index;
+  unordered_map<string, unique_ptr<Node>> forest;
+
+  Status s = ConstructForest(raft_config, proxy_topology, &index, &forest);
+  if (PREDICT_FALSE(!s.ok() && !s.IsIncomplete())) {
+    return s;
+  }
+  RETURN_NOT_OK(MergeForestIntoSingleRoutingTree(leader_uuid, index, &forest));
+  ConstructNextHopIndicesRec(forest.begin()->second.get());
+
+  index_ = std::move(index);
+  topology_root_ = std::move(forest.begin()->second);
+
+  return s;
+}
+
+Status RoutingTable::ConstructForest(
+    const RaftConfigPB& raft_config,
+    const ProxyTopologyPB& proxy_topology,
+    std::unordered_map<std::string, Node*>* index,
+    std::unordered_map<std::string, std::unique_ptr<Node>>* forest) {
+
+  unordered_map<string, string> dest_to_proxy_from; // keyed by directed edge destination
+  for (const auto& edge : proxy_topology.proxy_edges()) {
+    if (ContainsKey(dest_to_proxy_from, edge.peer_uuid())) {
+      return Status::InvalidArgument("invalid proxy topology: duplicate destination uuid",
+                                     edge.peer_uuid());
+    }
+    InsertOrDie(&dest_to_proxy_from, edge.peer_uuid(), edge.proxy_from_uuid());
+  }
+
+  // Initially, construct a forest comprised of all peers in the Raft config,
+  // with no proxy_from relationships represented.
+  std::unordered_map<std::string, std::unique_ptr<Node>> tmp_forest;
+  std::unordered_map<std::string, Node*> tmp_index;
+  for (const RaftPeerPB& peer : raft_config.peers()) {
+    unique_ptr<Node> hop(new Node(peer));
+    tmp_index.emplace(peer.permanent_uuid(), hop.get());
+    auto result = tmp_forest.emplace(peer.permanent_uuid(), std::move(hop));
+    if (!result.second) {
+      return Status::InvalidArgument("invalid raft config: duplicate peer uuid",
+                                     peer.permanent_uuid());
+    }
+  }
+
+  // proxy_from nodes specified in ProxyTopologyPB that were not found in RaftConfigPB.
+  vector<string> proxy_from_nodes_not_found;
+
+  // Now, organize the forest into parent-child relationships, where the parent
+  // is represented as proxy_from in each ProxyTopologyPB edge, and the child
+  // is the destination. Any node without a valid proxy_from (either not
+  // specified in ProxyTopologyPB or specified as a peer that isn't currently a
+  // member of the Raft config) will be left as a tree root in the forest.
+  for (const RaftPeerPB& peer : raft_config.peers()) {
+    const string* proxy_from_uuid = FindOrNull(dest_to_proxy_from, peer.permanent_uuid());
+    if (!proxy_from_uuid || proxy_from_uuid->empty()) {
```

Comment:
nit: in a basic sanity check of ProxyTopologyPB we should disallow peers with empty UUID,
so that we don't keep having to check for empty uuid peers.

Reply:
Done